### PR TITLE
Pseudo code for topological sort

### PR DESCRIPTION
**Pull request by Ddiog-dev**
i used the pseudo code of the khan's algorithm found on wikipedia  and i added below the pseudo code a detailled example on how it works.
(https://en.wikipedia.org/wiki/Topological_sorting)
This PR concerns the issue #18

### CODE REVIEW

**Comment by ghost** [comment]
Although the pseudocode is correct, it does not go into the details of how to implement the above algorithm efficiently. (For example, it talks about deleting an edge which takes linear time, and hence the overall algorithm can jump upto **O(n^2)**.) However, in Kahn's Topological Sort, we don't need to delete the edge each time. The workaround for this is to delete a vertex virtually (effectively reducing the indegree). See [this](https://github.com/Just-A-Visitor/Algorithmic-Pseudocode/blob/9102ad0e74c66957d98b966fc4cde6bf4c7a9aaa/UVa%20Online%20Judge/Topological%20Sorting/Pickup%20Sticks%20(11686)/color_queue.cpp#L67) implementation for more details. (Line Number **67-99**)

Please make the required changes to the pseudocode. Also, do not place it inside the Sorting folder. Place it inside `Pseudocode/Graphs/Topological Sorting`.

Please do not include binary files (such as images) in your initial pull request. This is because they remain in the history even when you delete it (say, after the code review). This takes up the Repo space. If you want to attach binary files, include them only after the review is over. 

Let me know if you need any more clarifications

**Comment by Ddiog-dev** [comment]
Alright, no problems thank you for the feed back, i'm correctiong it right now !

**Comment by ghost** [comment]
Also, there's a typo in the name of the image. It's `Kahn's`

**Comment by Ddiog-dev** [comment]
So i moved the code in the right repositories, i deleted the images in the folder and i modified the pseudo code to match the edge handling in the code you gave me and i also clarified how to initialise the S set with inDegree

**Comment by ghost** [comment]
Here are a couple of suggestions to get you acquainted with our coding style

1) Don't use `camelCase` in LaTeX. (It looks weird with the rendering). Use `underscore_names`.

2) Enclose everything in a function. (See the new file for details on how to do this).

3) It's ok to leave a blank line in between the algorithm for clarity. Use `\Statex` for this purpose. See the new file to see it in usage.

4) As of now, your code describes what to do in plain English. However, the goal of this repo is to write compact pseudocodes. Hence, feel free to use concise statements to express the idea. Use comments to explain non-trivial parts. If you catch yourself writing long comments, explain it after the algorithm instead. (See the Pseudocode for `Median in a Stream of  Integers` for an example. In a nutshell, the pseudocode should be written in such a way that it can easily be converted to code in any language while at the same time acting as a language-independent explanation of the algorithm. See the pseudocodes of **CLRS** for motivation.

5) If you need to use a lot of spaces within math-mode, consider using the `\textit` environment instead.

6) If you wish, you can use the `Ensure` and `Require` field to specify constraints on the input. (See the Gist).

7) Avoid using `++` as the rendering is not good. (The plus becomes too big).

8) Include occasional comments using `\Comment`.

9) Use Macros such as `not empty`, `is`, etc. 

10) I personally prefer `For each` instead of `For all`

11) I've used the `hyperref` package for coloured links.

**Comment by ghost** [comment]
1) Please go through the above reviews and make sure that you understand each point.

2) I might have missed a lot of things. You can look out for the others in the new file. [Link](https://gist.github.com/Just-A-Visitor/92efda80f0ad0969c1a256ea1369bc61). Please go through each line and understand the changes made.

3) After you've gone through the file and understood the changes, overwrite your TEX file with the new file.

4) Generate a PDF (of the new file) using Overleaf. Name it `Topological Sorting.pdf`. Place it inside the same folder as `SourceCode.tex`

5) Inside the folder where you've placed the source code, create a new folder called `Images` (notice the capitalization). Upload all your images there.

6) I've also mentioned some reviews. Read them and click on resolve conversation.

6) Ping me when you are done.

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+\begin{document}
+
+\begin{algorithm}
+
+  \caption{Topological sort using Kahn's algorithm  \cite{1}}
+  
+  \begin{algorithmic}[1]
+    \Statex
+        \State $L \gets Empty\ list\ that\ will\ contain\ the\ sorted\ elements $
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: Use italics instead. (But I don't think it's required here)

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        \State $S \gets Set\ of\ all\ nodes\ with\ no\ incoming\ edge\ $
+        \State $inDegree \gets Vector\ where\ inDegree[n]\ is\ the\ number\ of\ incoming\ edge\ of\ the\ node\ n$
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: Don't use came case

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        \ForAll{$node\ n\ in\ the\ graph\ $}
+            \ForAll{$node\ m\ with\ an\ edge\ e\ from\ n\ to\ m$}
+                \State $inDegree[m]++\  (number\ of\ incoming\ edges\ of\ the\ node\ m) $
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: Don't use `++`

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+  \caption{Topological sort using Kahn's algorithm  \cite{1}}
+  
+  \begin{algorithmic}[1]
+    \Statex
+        \State $L \gets Empty\ list\ that\ will\ contain\ the\ sorted\ elements $
+        \State $S \gets Set\ of\ all\ nodes\ with\ no\ incoming\ edge\ $
+        \State $inDegree \gets Vector\ where\ inDegree[n]\ is\ the\ number\ of\ incoming\ edge\ of\ the\ node\ n$
+        \ForAll{$node\ n\ in\ the\ graph\ $}
+            \ForAll{$node\ m\ with\ an\ edge\ e\ from\ n\ to\ m$}
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: Be more concise

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                \State $inDegree[m]++\  (number\ of\ incoming\ edges\ of\ the\ node\ m) $
+            \EndFor
+        \EndFor
+        
+        \ForAll{$node\ n\ in\ the\ graph\ $}
+            \If{$inDegree[n]\ == 0$}
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: I prefer `is` to `==`

File: Pseudocode/Graphs/Topological Sort/SourceCode.tex [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                \State $ add\ n\ to\ S $
+                \EndIf
+        \EndFor
+        
+        
+        \While {$ S\ is\ non\ empty\  $}
```
Comment (ghost) [inline Pseudocode/Graphs/Topological Sort/SourceCode.tex]: It's better to give meaningful names such as `queue`

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by Ddiog-dev** [comment]
I did the modifications and i added the pdf , tell me when i can push the pictures for the example !

**Comment by ghost** [comment]
There are a lot of small details still missing (like Insertion in queue using `.push` rather than plain English, `for each` rather than `for  all `, use of adjacency list instead of vaguely talking about edges, .use of underscores rather than camelCase, initialization of indegree array, syntax highlighting of `is not empty`, typo in the last `if`, etc.


Since I've already dragged this too much, I'd request you to overwrite your TEX file with my version. You can find it [here](https://gist.github.com/Just-A-Visitor/92efda80f0ad0969c1a256ea1369bc61). And by overwriting, I don't mean Modify your code, just copy paste my code and replace it. I think there was a confusion last time. (I usually provide the full review in the code itself, but since I'm a bit occupied, I decided to create a new version so that you can see the differences yourself)

When you are done, let me know so that I can allow you to upload the final pdf and images.

**Comment by Ddiog-dev** [comment]
Changes done

**Comment by ghost** [comment]
Looks good. You can upload the images now. (Create another folder `Images` inside the directory where sourceCode.tex is placed and place all the images there)

**Comment by Ddiog-dev** [comment]
Done

**Comment by ghost** [comment]
Great! Thank you for your contribution to this repository. I hope you got to learn something new from this experience.
